src/Polar/kernels.jl

```diff
@@ -204,31 +205,92 @@ KA.@kernel function adj_residual_node_kernel!(F, adj_F, vm, adj_vm, va, adj_va,
     end
 end
 
+KA.@kernel function adj_residual_node_kernel1!(adj_vm, adj_va,
+                                  colptr, rowval,
+                                  edge_vm_from, edge_va_from) 
+    i = @index(Global, Linear)
+    @inbounds for c in colptr[i]:colptr[i+1]-1
+        adj_vm[i] += edge_vm_from[c]
+        adj_va[i] += edge_va_from[c]
+    end
+end
+
+KA.@kernel function adj_residual_node_kernel2!(adj_vm, adj_va, 
+                                   edge_vm_a, edge_vm_i, edge_vm_j,
+                                   edge_va_a, edge_va_i, edge_va_j,
+)
+    nnz = length(edge_vm_a)
+    i = @index(Global, Linear)
+    for c in 1:nnz
+        if edge_va_j[c] == i
+            adj_vm[i] += edge_vm_a[c]
+            adj_va[i] += edge_va_a[c]
+        end
+    end
+end
 function adj_residual_polar!(
-    F, adj_F, vm, adj_vm, va, adj_va,
+    F::CuVector{T}, adj_F::CuVector{T}, vm, adj_vm, va, adj_va,
     ybus_re, ybus_im,
     pinj, adj_pinj, qinj, pv, pq, nbus
-)
+) where {T}
     npv = length(pv)
     npq = length(pq)
     nvbus = length(vm)
     nnz = length(ybus_re.nzval)
-    T = eltype(vm)
+    colptr = ybus_re.colptr
+    rowval = ybus_re.rowval
+    edge_vm_from = CUDA.zeros(T, nnz)
+    edge_vm_to   = CUDA.zeros(T, nnz)
+    edge_va_from = CUDA.zeros(T, nnz)
+    edge_va_to   = CUDA.zeros(T, nnz)
+    kernel_edge! = adj_residual_edge_kernel!(KA.CUDADevice())
+    kernel_node1! = adj_residual_node_kernel1!(KA.CUDADevice())
+    kernel_node2! = adj_residual_node_kernel2!(KA.CUDADevice())
+    spedge_vm_to = CUSPARSE.CuSparseMatrixCSR{T}(colptr, rowval, edge_vm_to,(nvbus,nvbus))
+    spedge_va_to = CUSPARSE.CuSparseMatrixCSR{T}(colptr, rowval, edge_va_to,(nvbus,nvbus))
+    tspedge_vm_to = CUSPARSE.CuSparseMatrixCOO(spedge_vm_to) 
+    tspedge_va_to = CUSPARSE.CuSparseMatrixCOO(spedge_va_to) 
+    ev = kernel_edge!(F, adj_F, vm, adj_vm, va, adj_va,
+                 ybus_re.colptr, ybus_re.rowval,
+                 ybus_re.nzval, ybus_im.nzval,
+                 edge_vm_from, edge_vm_to,
+                 edge_va_from, edge_va_to,
+                 pinj, adj_pinj, qinj, pv, pq,
+                 ndrange=npv+npq)
+    wait(ev)
+    tspedge_vm_to.nzVal .= spedge_vm_to.nzVal
+    tspedge_va_to.nzVal .= spedge_va_to.nzVal
+    ev = kernel_node1!(adj_vm, adj_va,
+            ybus_re.colptr, ybus_re.rowval,
+            edge_vm_from, edge_va_from, 
+            ndrange=nvbus)
+    wait(ev)
+    ev = kernel_node2!(adj_vm, adj_va,
+            tspedge_vm_to.nzVal, tspedge_vm_to.rowInd, tspedge_vm_to.colInd,
+            tspedge_va_to.nzVal, tspedge_va_to.rowInd, tspedge_va_to.colInd,
+            ndrange=nnz)
+    wait(ev)
```

Comment:
Do you know whether we could merge `kernel_node1!` and `kernel_node2!` ?

Reply:
Could be. Let's look at that now.